plugins/hls-cabal-fmt-plugin/src/Ide/Plugin/CabalFmt.hs

```diff
@@ -0,0 +1,66 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Ide.Plugin.CabalFmt where
+
+import Control.Lens
+import Control.Monad.IO.Class
+import qualified Data.Text as T
+import Development.IDE hiding (pluginHandlers)
+import Ide.PluginUtils
+import Ide.Types
+import Language.LSP.Types as J
+import qualified Language.LSP.Types.Lens as J
+import System.Exit
+import System.FilePath
+import System.Process
+import Prelude hiding (log)
+import System.Directory
+
+data Log
+  = LogProcessInvocationFailure Int
+  | LogReadCreateProcessError String [String]
+  deriving (Show)
+
+instance Pretty Log where
+  pretty = \case
+    LogProcessInvocationFailure code -> pretty ("Invocation of cabal-fmt failed with code" :: String) <+> pretty code
+    LogReadCreateProcessError errorMsg args -> pretty ("Invocation of cabal-fmt" :: String) <+> pretty errorMsg <+> pretty ("with arguments" :: String) <+> pretty args
+
+descriptor :: Recorder (WithPriority Log) -> PluginId -> PluginDescriptor IdeState
+descriptor recorder plId =
+  (defaultCabalPluginDescriptor plId)
+    { pluginHandlers = mkFormattingHandlers (provider recorder)
+    }
+
+-- | Formatter provider of cabal fmt.
+-- Formats the given source in either a given Range or the whole Document.
+-- If the provider fails an error is returned that can be displayed to the user.
+provider :: Recorder (WithPriority Log) -> FormattingHandler IdeState
+provider recorder _ide _typ contents nfp opts = liftIO $ do
+  let cabalFmtArgs = [fp, "--indent", show tabularSize] ++ ["--tabular" | tabs]
+  x <- findExecutable "cabal-fmt"
+  case x of
+    Just _ -> do
+      (exitCode, out, err) <-
+        readCreateProcessWithExitCode
+          ( proc "cabal-fmt" cabalFmtArgs
+          )
+            { cwd = Just $ takeDirectory fp
+            }
+          ""
+      log Debug $ LogReadCreateProcessError err cabalFmtArgs
+      case exitCode of
+        ExitFailure code -> do
+          log Warning $ LogProcessInvocationFailure code
+          pure $ Left (ResponseError UnknownErrorCode "Failed to invoke cabal-fmt" Nothing)
```

Comment:
```suggestion
          throwPluginError cabalFmtId "Failed to invoke cabal-fmt" "provider"
```

Reply:
This holds for the other `Nothing` cases as well

Reply:
Nevermind this won't work in your current setup. Food for thought though :)

Reply:
It's sort of silly to go through the mechanism to throw the error and then unwrap it directly into a `ResponseError` but idk.